rfcs/20220329-Extending-CUDA-backend-of-oneDNN-Primitives/README.md

```diff
@@ -0,0 +1,190 @@
+
+- Similarly, the functionality of `reduction_norm_lp_max` and `reduction_norm_lp_sum` algortihms can be implemented using `CUDNN_REDUCE_TENSOR_NORM1` algorithm mode of `cudnnReduceTensor` in combination with `CUDNN_REDUCE_TENSOR_MAX` and `CUDNN_REDUCE_TENSOR_ADD` respectively. 
+
+    - However, this workaround doesn't provide the flexibility to use arbitrary value for power `p` other than 2.  
+
+- As an alternative to the above workarounds, the functionality of the above algorithms can be implemented through a custom kernel in SYCL. 
+
+ 
+
+### Recurrent Neural Networks 
+
+- RNN primitive in CUDA backend can be implemented using `cudnnRNNForward` and `cudnnRNNBackward` primitives of cuDNN. 
+
+- cuDNN's primitives only support two algorithm of RNN. 
+
+    - `CUDNN_LSTM` 
+
+    - `CUDNN_GRU` 
+
+- Supported Data Types: f16, f32 
+
+ 
+
+#### Propagation 
+
+Supported propagation kinds: Forward Training/Inference and Backward 
+
+ 
+
+#### Implementation Limitations 
+
+- No support for following algorithms in cuDNN. 
+
+    - `augru` 
+
+    - `lbr_augru` 
+
+    - `lbr_gru` 
+
+    - `vanilla_rnn` 
+
+- No support for U8, bf16 datatypes. 
+
+ 
+
+#### Proposed Solution 
+
+The unsupported algorithms of `augru`, `lbr_augru`, `lbr_gru`, `vanilla_rnn` in cuDNN can be implemented using custom kernels, per the mathematical expressions of the algorithms. 
+
+ 
+
+### PRelu  
+
+There is no equivalent primitive API in cuDNN for PRelu. 
+
+ 
+
+#### Propagation 
+
+Forward Training/Inference and Backward 
+
+ 
+
+#### Implementation Limitations 
+
+Lack of equivalent support for PRelu primitive in cuDNN. 
+
+ 
+
+#### Proposed Solution 
+
+- PRelu primitive can be implemented using cuDNN with `cudnnActivationForward()` and `cudnnActivationBackward()`. 
+
+    - Using `CUDNN_ACTIVATION_RELU` algorithm mode, it is possible to implement LEAKY RELU with a constant scalar alpha value, but not PRelu with a tensor alpha value. 
+
+    - However, the functionality of PRelu can be accomplished using `CUDNN_ACTIVATION_RELU` algorithm mode in combination with `cudnnOpTensor()`.
+
+- As a better efficient alternative, the primitive could be implemented using custom kernel, emulating the mathematical functionality of PRelu. 
+
+ 
+
+### Layer Normalization 
+
+There is no equivalent primitive API In cuDNN for Layer Normalization. 
+
+ 
+
+#### Propagation 
+
+Forward Training/Inference, Backward and Backward_Data 
+
+ 
+
+#### Implementation Limitations 
+
+Lack of equivalent support for Layer Normalization primitive in cuDNN. 
+
+ 
+
+#### Proposed Solution 
+
+- Layer Normalization primitive can be implemented using cuDNN with `cudnnBatchNormalizationForwardInference`, `cudnnBatchNormalizationForwardTraining` and `cudnnBatchNormalizationBackward` primitive APIs.  
+
+    - This can be achieved with the algorithm mode of `CUDNN_BATCHNORM_PER_ACTIVATION` and by limiting the batch size to 1. 
```

Comment:
The layer normalization was introduced mostly for RNN therefore it support a specific case where normalization is performed over the last axis.

Can the cuDNN API for batch normalization allow us to do that or are you going to do some data manipulation prior to calling the API?

Also, layer normalization also supports the following flags:

- dnnl_use_global_stats
- dnnl_use_scaleshift (will probably be deperecated)
- dnnl_use_scale
- dnnl_use_shift